processing/src/main/java/io/druid/query/groupby/GroupByQuery.java

```diff
@@ -251,46 +260,81 @@ public int compare(Object lhs, Object rhs)
 
   public Ordering<Row> getRowOrdering(final boolean granular)
   {
+    final boolean sortByDimsFirst = getContextSortByDimsFirst();
+
     final Comparator naturalNullsFirst = Ordering.natural().nullsFirst();
 
-    return Ordering.from(
-        new Comparator<Row>()
-        {
-          @Override
-          public int compare(Row lhs, Row rhs)
+    if (sortByDimsFirst) {
+      return Ordering.from(
+          new Comparator<Row>()
           {
-            final int timeCompare;
+            @Override
+            public int compare(Row lhs, Row rhs)
+            {
+              for (DimensionSpec dimension : dimensions) {
+                final int dimCompare = naturalNullsFirst.compare(
+                    lhs.getRaw(dimension.getOutputName()),
+                    rhs.getRaw(dimension.getOutputName())
+                );
+                if (dimCompare != 0) {
+                  return dimCompare;
+                }
+              }
 
-            if (granular) {
-              timeCompare = Longs.compare(
-                  granularity.truncate(lhs.getTimestampFromEpoch()),
-                  granularity.truncate(rhs.getTimestampFromEpoch())
-              );
-            } else {
-              timeCompare = Longs.compare(
-                  lhs.getTimestampFromEpoch(),
-                  rhs.getTimestampFromEpoch()
-              );
+              if (granular) {
```

Comment:
Maybe extract this + the next 10 lines into a method, which could be used in both comparators, for DRY and self-documenting?